File: 백준/Silver/1012. 유기농 배추/유기농 배추.py
import sys


def bfs(i, j):
    q = []
    q.append((i, j))
    visited[i][j] = 1
    arr[i][j] = 0
    while q:
        i, j = q.pop(0)
        for di, dj in [(0, 1), (1, 0), (0, -1), (-1, 0)]:
            ni = i + di
            nj = j + dj
            if 0 <= ni < N and 0 <= nj < M and arr[ni][nj] == 1 and visited[ni][nj] == 0:
                q.append((ni, nj))
                visited[ni][nj] = 1
                arr[ni][nj] = 0


T = int(sys.stdin.readline())
for _ in range(1, T+1):
    M, N, K = map(int, sys.stdin.readline().split())
    arr = [[0] * M for _ in range(N)]
    visited = [[0] * M for _ in range(N)]
    for _ in range(K):
        Y, X = map(int, sys.stdin.readline().split())
        arr[X][Y] = 1

    ans = 0
    for x in range(N):
        for y in range(M):
            if arr[x][y] == 1:
                bfs(x, y)
                ans += 1

    print(ans)

File: 백준/Silver/1012. 유기농 배추/유기농 배추.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {

	static int M, N, K;
	static int[][] arr;
	static boolean[][] visited;
	static int[] dx = { 0, 1, 0, -1 };
	static int[] dy = { 1, 0, -1, 0 };

	public static void main(String[] args) throws NumberFormatException, IOException {

		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int T = Integer.parseInt(br.readLine());
		for (int t = 0; t < T; t++) {
			StringTokenizer st = new StringTokenizer(br.readLine());

			N = Integer.parseInt(st.nextToken());
			M = Integer.parseInt(st.nextToken());
			K = Integer.parseInt(st.nextToken());

			arr = new int[N][M];

			int answer = 0;
			for (int i = 0; i < K; i++) {
				st = new StringTokenizer(br.readLine());
				int x = Integer.parseInt(st.nextToken());
				int y = Integer.parseInt(st.nextToken());
				arr[x][y] = 1;
			}

			visited = new boolean[N][M];
			for (int i = 0; i < N; i++) {
				for (int j = 0; j < M; j++) {
					if (arr[i][j] == 1 && !visited[i][j]) {
						answer++;
						dfs(i, j);
					}
				}
			}
			System.out.println(answer);
		}
	}

	static void dfs(int x, int y) {
		visited[x][y] = true;

		for (int d = 0; d < 4; d++) {
			int nx = x + dx[d], ny = y + dy[d];
			if (0 > nx || N <= nx || 0 > ny || M <= ny)
				continue;
			if (visited[nx][ny])
				continue;
			if (arr[nx][ny] == 1) {
				dfs(nx, ny);
			}
		}
	}
}
